segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/cache/DirectMemoryBuffer.java

```diff
@@ -0,0 +1,493 @@
+
+    @Override
+    public synchronized void close() {
+        if (this.buf != null && this.buf.refCnt() > 0) {
+            this.buf.release();
+            this.buf = null;
+        }
+
+        this.usedBlockCount = -1;
+    }
+
+    //endregion
+
+    //region Properties
+
+    /**
+     * Gets a value indicating the number of used Blocks in this buffer.
+     *
+     * @return The number of used blocks.
+     */
+    synchronized int getUsedBlockCount() {
+        return this.usedBlockCount;
+    }
+
+    /**
+     * Gets a value indicating whether this buffer has allocated any memory yet.
+     *
+     * @return The result.
+     */
+    synchronized boolean isAllocated() {
+        return this.buf != null;
+    }
+
+    /**
+     * Gets a value indicating whether this buffer has any capacity to add new blocks (NOTE: there may be capacity for
+     * appending to existing blocks, but this method refers to new blocks).
+     *
+     * @return True if at least one block is free, false otherwise.
+     */
+    synchronized boolean hasCapacity() {
+        return this.usedBlockCount > 0 && this.usedBlockCount < this.layout.blocksPerBuffer();
+    }
+
+    @Override
+    public synchronized String toString() {
+        return String.format("Id=%d, UsedBlockCount=%d", this.id, this.usedBlockCount);
+    }
+
+    //endregion
+
+    //region Buffer Implementation
+
+    /**
+     * Writes new data to this buffer, if {@link #hasCapacity()} is true.
+     *
+     * This will begin writing data to the first available block in the buffer, and filling blocks until either running
+     * out of available blocks or all the data has been written.
+     *
+     * @param data               An {@link BufferView} representing the data to add. The {@link BufferView#copyTo(ByteBuffer)}
+     *                           method will be used to transfer the data to the appropriate Buffer Block.
+     * @param predecessorAddress The address of the Buffer-Block that precedes the write to this Buffer. If none, this
+     *                           should be set to {@link  CacheLayout#NO_ADDRESS}.
+     * @return A {@link WriteResult} representing the result of the write, or null if {@link #hasCapacity()} is false.
+     * If all the data has been written, then  {@link WriteResult#getRemainingLength()} will be 0.
+     */
+    synchronized WriteResult write(BufferView data, int predecessorAddress) {
+        if (this.usedBlockCount >= this.layout.blocksPerBuffer()) {
+            // Full
+            return null;
+        }
+
+        ByteBuf metadataBuf = getMetadataBlock();
+        long blockMetadata = metadataBuf.getLong(0);
+        int blockId = this.layout.getNextFreeBlockId(blockMetadata);
+        assert blockId != CacheLayout.NO_BLOCK_ID;
+        final int firstBlockId = blockId;
+        int dataOffset = 0;
+        try {
+            // Use a do-while loop since we want to properly handle the (valid) case when data.getLength() == 0.
```

Comment:
Well, we can very well insert a zero-length entry in the Cache Storage. That is a bad idea indeed, but the calling code may want to do this and the Cache Storage needs to support this case. The only way to make it work is to allocate a whole block for it so that we know it's registered. 

As of now there is no case in the Segment Store where it would insert a zero-length entry, but this is a possibility so we need to handle it.